text/000-wasm-bindgen-trait-method-inheritance.md

```diff
@@ -0,0 +1,384 @@
+re-exports all of the traits. So that way the user can just put this at the top of
+their module:
+
+```rust
+// Now all of the methods work!
+use web_sys::traits::*;
+```
+
+# Rationale and Alternatives
+[alternatives]: #rationale-and-alternatives
+
+There is essentially only one other alternative design: inherit impl.
+
+Rather than using traits, it can instead use inherit impls on every type in the class inheritance hierarchy.
+
+As an example, the WebIDL generator could generate this code for the `EventTarget` and `Node` types:
+
+```rust
+#[wasm_bindgen]
+extern {
+    type EventTarget;
+
+    // Methods from EventTarget
+    #[wasm_bindgen(method, js_name = addEventListener)]
+    pub fn add_event_listener(this: &EventTarget, type: &str, callback: &Closure<FnMut(Event)>, options: bool);
+
+    #[wasm_bindgen(method, js_name = removeEventListener)]
+    pub fn remove_event_listener(this: &EventTarget, type: &str, callback: &Closure<FnMut(Event)>, options: bool);
+
+    #[wasm_bindgen(method, js_name = dispatchEvent)]
+    pub fn dispatch_event(this: &EventTarget, event: Event);
+}
+```
+
+```rust
+#[wasm_bindgen]
+extern {
+    type Node;
+
+
+    // Methods from EventTarget
+    #[wasm_bindgen(method, js_name = addEventListener)]
+    pub fn add_event_listener(this: &Node, type: &str, callback: &Closure<FnMut(Event)>, options: bool);
+
+    #[wasm_bindgen(method, js_name = removeEventListener)]
+    pub fn remove_event_listener(this: &Node, type: &str, callback: &Closure<FnMut(Event)>, options: bool);
+
+    #[wasm_bindgen(method, js_name = dispatchEvent)]
+    pub fn dispatch_event(this: &Node, event: Event);
+
+
+    // Methods from Node
+    #[wasm_bindgen(method, getter = nodeName)]
+    pub fn node_value(this: &Node) -> JsString;
+
+    #[wasm_bindgen(method, getter = textContent)]
+    pub fn text_content(this: &Node) -> JsString;
+
+    #[wasm_bindgen(method, js_name = appendChild)]
+    pub fn append_child(this: &Node, node: Node) -> Node;
+
+    #[wasm_bindgen(method, js_name = removeChild)]
+    pub fn remove_child(this: &Node, child: Node) -> Node;
+}
+```
+
+As you can see, it duplicates the `add_event_listener`, `remove_event_listener`, and `dispatch_event`
+methods on `Node`.
+
+Similarly, it would have to duplicate all of the `EventTarget` and `Node` methods on `Element`. And it
+would have to duplicate all of the `EventTarget`, `Node`, and `Element` methods on `HTMLElement`, etc.
+
+This is an incredible amount of duplication, so it's only really feasible for a tool which automatically
+generates the methods (such as the WebIDL generator). Trying to do this duplication by hand is unmaintainable.
+
+That means that if you're using inherit impls, it will be very painful to use method inheritance with anything
+other than WebIDL, because of the maintenance burden.
+
+And because class-based inheritance is used outside of WebIDL, we want to be able to support non-WebIDL use
+cases.
```

Comment:
One other alternative is to use `Deref` to model single inheritance. This doesn't suffer from the same drawbacks mentioned here, but it is generally considered an [anti-pattern](https://github.com/rust-unofficial/patterns/blob/master/anti_patterns/deref.md).